ir/CMakeLists.txt

```diff
@@ -64,6 +64,7 @@ set (BASE_IR_DEF_FILES
 set (IR_DEF_FILES ${IR_DEF_FILES} ${BASE_IR_DEF_FILES} PARENT_SCOPE)
 
 add_library (ir STATIC ${IR_SRCS})
+target_link_libraries(ir absl::flat_hash_map)
```

Comment:
Not sure how we want to manage linking... Should each dependent target link absl::flat_hash_map, should `ir` export it? Also unclear to me what the best approach is to keep binary size small. 
I am guess we should make `PUBLIC` explicit here.

Reply:
I'll revise this PR wrt dependencies

Reply:
`add_dependency` was used because historically all the libs were built independently and linked together by each back end. However, to ensure proper build order in parallel builds, `add_dependency` was needed. 

The fine-grained approach works for me. #4474 implements some of this as I ran into some issues downstream.

Reply:
> `add_dependency` was used because historically all the libs were built independently and linked together by each back end. However, to ensure proper build order in parallel builds, `add_dependency` was needed.

Yes. But this is not how the dependencies should be done. Things are pulling headers, so this is a build-time dependency, not a link-time...

Reply:
With binary size I meant the final compiler binaries, which can be huge. I believe this is primarily caused by all the templating and ir-generated header, but also by the chaotic linking, which also seems to affect linking time. It can be excruciatingly slow.

Reply:
> Yes. But this is not how the dependencies should be done. Things are pulling headers, so this is a build-time dependency, not a link-time... #4473 also has some... we'd need to unify all somehow

Yeah I agree with you, the compiler infrastructure has been going through various phases of structuring with each back end doing their own thing.